cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java

```diff
@@ -140,97 +155,205 @@ public static void finish() {
+    Files.copy(bobKeytabFile, Locations.newOutputSupplier(bobRemoteKeytabFile));
+
+    OwnerAdmin ownerAdmin = getOwnerAdmin();
+
+    // add an owner for stream and dataset
+    StreamId aliceEntity = new StreamId("ugiProviderNs", "dummyStream");
+    namespaceClient.create(new NamespaceMeta.Builder().setName("ugiProviderNs").build());
+    ownerAdmin.add(aliceEntity, aliceKerberosPrincipalId);
+    DatasetId bobEntity = new DatasetId("ugiProviderNs", "dummyDataset");
+    ownerAdmin.add(bobEntity, bobKerberosPrincipalId);
+
+    DefaultUGIProvider provider = new DefaultUGIProvider(cConf, locationFactory, ownerAdmin, namespaceClient);
+
+    ImpersonationRequest aliceImpRequest = new ImpersonationRequest(aliceEntity, ImpersonatedOpType.OTHER);
+    UGIWithPrincipal aliceUGIWithPrincipal = provider.getConfiguredUGI(aliceImpRequest);
+    Assert.assertEquals(UserGroupInformation.AuthenticationMethod.KERBEROS,
+                        aliceUGIWithPrincipal.getUGI().getAuthenticationMethod());
+    Assert.assertTrue(aliceUGIWithPrincipal.getUGI().hasKerberosCredentials());
+    Assert.assertEquals(aliceKerberosPrincipalId.getPrincipal(), aliceUGIWithPrincipal.getPrincipal());
+
+    // Fetch it again, it is should return the same UGI since there is caching
+    Assert.assertSame(aliceUGIWithPrincipal.getUGI(), provider.getConfiguredUGI(aliceImpRequest).getUGI());
+
+    ImpersonationRequest bobImpRequest = new ImpersonationRequest(bobEntity, ImpersonatedOpType.OTHER);
+    UGIWithPrincipal bobUGIWithPrincipal = provider.getConfiguredUGI(bobImpRequest);
+    Assert.assertEquals(UserGroupInformation.AuthenticationMethod.KERBEROS,
+                        bobUGIWithPrincipal.getUGI().getAuthenticationMethod());
+    Assert.assertTrue(bobUGIWithPrincipal.getUGI().hasKerberosCredentials());
+    Assert.assertEquals(bobKerberosPrincipalId.getPrincipal(), bobUGIWithPrincipal.getPrincipal());
+
+    Assert.assertTrue(bobRemoteKeytabFile.delete());
+
+    // Fetch the bob UGI again, it should still return the valid one
+    Assert.assertSame(bobUGIWithPrincipal, provider.getConfiguredUGI(bobImpRequest));
+
+    // Invalid the cache, getting of Alice UGI should pass, while getting of Bob should fails
+    provider.invalidCache();
+    Assert.assertNotSame(aliceUGIWithPrincipal, provider.getConfiguredUGI(aliceImpRequest));
+    try {
+      provider.getConfiguredUGI(bobImpRequest);
+      Assert.fail("Expected IOException when getting UGI for " + bobImpRequest);
+    } catch (IOException e) {
+      // Expected
+    }
+
+    // cleanup
+    ownerAdmin.delete(aliceEntity);
+    ownerAdmin.delete(bobEntity);
+    namespaceClient.delete(new NamespaceId("ugiProviderNs"));
+  }
+
+  @Test
+  public void testRemoteUGIProvider() throws Exception {
+    // Starts a mock server to handle remote UGI requests
+    final NettyHttpService httpService = NettyHttpService.builder("remoteUGITest")
+      .addHttpHandlers(Collections.singleton(new UGIProviderTestHandler()))
```

Comment:
It is testing the RemoteUGI provider. If you see the UGI provider which is created is Remote. Only the handler is mock and not the actual ImpersonationHandler. Its not testing the ImpersonationHandelr logic. I am not completely sure as why it doesn't use the actual ImpersonationHandler. I think the idea is not to start app-fabric here and test the just the RemoteUGIProvider.